plugins/fluent-bit/src/plugin.cpp

```diff
@@ -0,0 +1,505 @@
+  }
+
+  auto start() -> bool {
+    TENZIR_ASSERT_CHEAP(ctx_ != nullptr);
+    TENZIR_DEBUG("starting Fluent Bit engine");
+    auto ret = flb_start(ctx_);
+    if (ret == 0)
+      return true;
+    TENZIR_ERROR("failed to start engine ({})", ret);
+    return false;
+  }
+
+  /// Stops the engine.
+  auto stop() -> bool {
+    TENZIR_ASSERT_CHEAP(ctx_ != nullptr);
+    TENZIR_DEBUG("stopping Fluent Bit engine");
+    for (size_t i = 0; ctx_->status == FLB_LIB_OK && i < num_stop_polls_; ++i) {
+      TENZIR_DEBUG("sleeping while Fluent Bit context is okay");
+      std::this_thread::sleep_for(poll_interval_);
+    }
+    auto ret = flb_stop(ctx_);
+    if (ret == 0)
+      return true;
+    TENZIR_ERROR("failed to stop engine ({})", ret);
+    return false;
+  }
+
+  flb_ctx_t* ctx_{nullptr}; ///< Fluent Bit context
+  int ffd_{-1};             ///< Fluent Bit handle for pushing data
+  std::chrono::milliseconds poll_interval_{}; ///< How fast we check FB
+  size_t num_stop_polls_{0}; ///< Number of polls in the destructor
+  std::string buffer_{};     ///< Buffer shared with Fluent Bit
+  std::unique_ptr<std::mutex> buffer_mtx_{}; ///< Protects the shared buffer
+};
+
+class fluent_bit_operator final : public crtp_operator<fluent_bit_operator> {
+public:
+  fluent_bit_operator() = default;
+
+  fluent_bit_operator(operator_args args, record config)
+    : args_{std::move(args)}, config_{std::move(config)} {
+  }
+
+  auto operator()(operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    auto engine = engine::make_source(args_, config_);
+    if (not engine) {
+      diagnostic::error("failed to create Fluent Bit engine")
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+    auto builder = adaptive_table_slice_builder{};
+    auto parse = [&builder](std::string_view data) {
+      TENZIR_DEBUG("processing shared buffer");
+      TENZIR_ASSERT(not data.empty());
+      // What we're getting here is the typical Fluent Bit array consisting of
+      // the following format, as described in
+      // https://docs.fluentbit.io/manual/concepts/key-concepts#event-format:
+      //
+      //     [[TIMESTAMP, METADATA], MESSAGE]
+      //
+      // where
+      //
+      // - TIMESTAMP is a timestamp in seconds as an integer or floating point
+      //   value (not a string);
+      // - METADATA is a possibly-empty object containing event metadata; and
+      // - MESSAGE is an object containing the event body.
+      //
+      // Fluent Bit versions prior to v2.1.0 instead used
+      //
+      //     [TIMESTAMP, MESSAGE]
+      //
+      // to represent events. This format is still supported for reading input
+      // event streams.
+      //
+      // We are parsing this into a table with the following fields:
+      //
+      // 1. timestamp: time (timestamp alias type)
+      // 2. metadata: record (inferred)
+      // 3. message: record (inferred)
+      //
+      // TODO: do 👆 as opposed to doing 👇
```

Comment:
@jachris I completed the table slice building to the point where two things are missing:

1. Replacing the old adaptive builder with the new one
2. Using simdjson to parse the above structure, i.e., accept both `[[TIMESTAMP, METADATA], MESSAGE]` and `[TIMESTAMP, MESSAGE]`. These are new-line delimited *array*, but that's valid JSONL.

I have neither worked with the API of (1) or (2), and would therefore be thankful if you  could help me with that last bit.

Reply:
I've used the new series builder successfully to do this. @jachris beautiful API!